types/json.go

```diff
@@ -104,22 +107,28 @@ func (a JSONArray) String() string {
 		}
 		switch {
 		case a.IsNull(i):
-			o.WriteString("(null)")
+			o.WriteString(array.NullValueStr)
 		default:
-			fmt.Fprintf(o, "\"%s\"", arr.Value(i))
+			fmt.Fprintf(o, "%q", a.ValueStr(i))
 		}
 	}
 	o.WriteString("]")
 	return o.String()
 }
 
+func (a *JSONArray) Value(i int) json.RawMessage {
```

Comment:
TODO decided:
* `Value` -> `json.RawMessage`, `GetOneForMarshal` -> parsed value
* `Value` -> parsed value, `GetOneForMarshal` -> `json.RawMessage`
cc: @yevgenypats @hermanschaaf

Reply:
The JSON type was the most difficult one to deal with in the CSV and Parquet writers/readers, and I think it's because it's unclear what it represents, and so some of the methods are inconsistent. 

For example, should calling MarshalJSON on the JSON type give you unescaped JSON, or an escaped JSON string? When you append values, should you append strings that are already marshalled JSON, or the raw objects?

In other words, does it represent any Go object (which happens to be stored on disk as JSON), or does it represent a valid JSON string?

I think it doesn't matter what we decide, we just need to pick one and stick with it. The other implementations can then know what to expect and handle it accordingly.

If I have to pick one, I'd vote for treating everything like a string, so callers need to call json.Marshal before sending it to the builder, and Value() will just return a string. The main purpose of this type will then be to document that the stored string contains valid JSON, which can be useful in some databases.

Reply:
Discussed internally, proposed plan:

* `Builder` implementation:
  * `Append(v any)` takes any value and marshals it (including `struct{...}`, `slice`, `nil` & `json.RawMessage`)
  * `AppendValues(v []any, valid []bool)` is same as `Append` with respect to `valid` slice
  * `AppendValueFromString(s string) error` will unmarshal to verify that the value is indeed the correct JSON, then save the `json.RawMessage`
  * `UnmarshalOne(dec *json.Decoder) error` will unmarshal & store using `Append` / `AppendNull`
  * **new**: `Unmarshal(dec *json.Decoder)` - unmarshal encoded slice
  * `UnmarshalJSON([]byte) error` - same as one, but accounts for value being marshaled slice
* `Array` implementation
  * `String() string` - joined of quoted `json.RawMessage` values cast to strings
  * `Value(i int) any` - unmarshaled value (`map`, `slice`, etc.)
  * `ValueStr(i int) string` - `string(json.RawMesaage)`, no quoting performed
  * `MarshalJSON() ([]byte, error)` - marshal as slice of `json.RawMessage`
  * `GetOneForMarshal(i int) any` - return stored `json.RawMessage`
```

The inspiration for the `Unmarshal` functions is taken from Apache Arrow builders

Reply:
Done in 7f6afc8